
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Executing scAVENGERS &#8212; scAVENGERS 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="executing-scavengers">
<h1>Executing scAVENGERS<a class="headerlink" href="#executing-scavengers" title="Permalink to this heading">¶</a></h1>
<section id="running-the-whole-pipeline">
<h2>Running the whole pipeline<a class="headerlink" href="#running-the-whole-pipeline" title="Permalink to this heading">¶</a></h2>
<section id="execution">
<h3>Execution<a class="headerlink" href="#execution" title="Permalink to this heading">¶</a></h3>
<p>The whole pipeline is executed by running the command below.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>conda activate scavengers
scAVENGERS pipeline --configfile config.yaml -j $THREADS
</pre></div>
</div>
</section>
<section id="preparing-data">
<h3>Preparing data<a class="headerlink" href="#preparing-data" title="Permalink to this heading">¶</a></h3>
<p>Alignment file in bam format and line-seperated list of barcodes in text file are required to run scAVENGERS.</p>
<p>If you are using products from 10X Genomics, you can acquire the alignment file and barcodes in ease by using Cellranger ATAC count (https://github.com/10XGenomics/cellranger-atac). The tool aligns fastq files by using bwa mem-like algorithm.</p>
<ul class="simple">
<li><p>The alignments are stored in <code class="docutils literal notranslate"><span class="pre">outs/possorted_bam.bam</span></code></p></li>
<li><p>The barcodes are stored in <code class="docutils literal notranslate"><span class="pre">outs/filtered_peak_bc_matrix/barcodes.tsv</span></code>.</p></li>
</ul>
</section>
<section id="configuring">
<h3>Configuring<a class="headerlink" href="#configuring" title="Permalink to this heading">¶</a></h3>
<p>The configuration file is provided in <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> in scAVENGERS repository.</p>
</section>
</section>
<section id="running-the-demultiplexing-module">
<h2>Running the demultiplexing module<a class="headerlink" href="#running-the-demultiplexing-module" title="Permalink to this heading">¶</a></h2>
<p>The demultiplexing module <code class="docutils literal notranslate"><span class="pre">scAVENGERS</span> <span class="pre">cluster</span></code> is executed by running the command below.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">activate</span> <span class="n">scavengers</span>
<span class="n">scAVENGERS</span> <span class="n">cluster</span> <span class="o">-</span><span class="n">a</span> <span class="n">alt</span><span class="o">.</span><span class="n">mtx</span> <span class="o">-</span><span class="n">r</span> <span class="n">ref</span><span class="o">.</span><span class="n">mtx</span> <span class="o">-</span><span class="n">b</span> <span class="n">barcodes</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">o</span> <span class="n">clusters</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
<section id="id1">
<h3>Preparing data<a class="headerlink" href="#id1" title="Permalink to this heading">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">scAVENGERS</span> <span class="pre">cluster</span></code> requires three inputs: count matrices for alternative and reference alleles in .mtx format, and a line-seperated list of barcodes in text file.</p>
<p>You can acquire allele count matrices by using VarTrix (https://github.com/10XGenomics/vartrix). Below is a minimal command to get reference and alternative allele counts.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vartrix</span> \
<span class="o">-</span><span class="n">b</span> <span class="n">alignment</span><span class="o">.</span><span class="n">bam</span> \
<span class="o">-</span><span class="n">v</span> <span class="n">variants</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="o">--</span><span class="n">fasta</span> <span class="n">reference</span><span class="o">.</span><span class="n">fa</span> \
<span class="o">-</span><span class="n">c</span> <span class="n">barcodes</span><span class="o">.</span><span class="n">txt</span> \
<span class="o">--</span><span class="n">scoring</span><span class="o">-</span><span class="n">method</span> <span class="n">coverage</span> \
<span class="o">--</span><span class="n">out</span><span class="o">-</span><span class="n">matrix</span> <span class="p">{</span><span class="n">output</span><span class="o">.</span><span class="n">alt</span><span class="p">}</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">matrix</span> <span class="p">{</span><span class="n">output</span><span class="o">.</span><span class="n">ref</span><span class="p">}</span> 
</pre></div>
</div>
</section>
<section id="id2">
<h3>Configuring<a class="headerlink" href="#id2" title="Permalink to this heading">¶</a></h3>
<p>You can set options from the command when you run scAVENGERS cluster.</p>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">scAVENGERS</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2022, Seungbeom Han.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/Running scAVENGERS/executing scAVENGERS.md.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>